geode-core/src/main/java/org/apache/geode/internal/cache/CacheDistributionAdvisor.java

```diff
@@ -142,17 +149,25 @@ public String toString() {
    *
    * @param excludeInRecovery if true then members in recovery are excluded
    */
-  private Set<InternalDistributedMember> adviseAllEventsOrCached(final boolean excludeInRecovery)
+  private Set<InternalDistributedMember> adviseAllEventsOrCached()
       throws IllegalStateException {
     getAdvisee().getCancelCriterion().checkCancelInProgress(null);
-    return adviseFilter(profile -> {
-      assert profile instanceof CacheProfile;
-      CacheProfile cp = (CacheProfile) profile;
-      if (excludeInRecovery && cp.inRecovery) {
-        return false;
-      }
-      return cp.cachedOrAllEventsWithListener();
-    });
+
+    // minimize volatile reads by copying ref to local var
+    long tempProfilesVersion = profilesVersion; // volatile read
+    if (adviseAllEventsVersion != tempProfilesVersion) {
+      adviseSetforAllEvents = Collections.unmodifiableSet(adviseFilter(profile -> {
+        assert profile instanceof CacheProfile;
```

Comment:
This new code on CacheDistributionAdvisor needs tests.
Add unit tests and since this code has some tricky concurrency issues add a ConcurrentTestRunner.
The ConcurrentTestRunner has been good at reproducing concurrency bugs.
See MaxLongGaugeConcurrentTest for an example of a simple ConcurrentTestRunner.